firebase-firestore/src/test/java/com/google/firebase/firestore/model/ObjectValueBuilderTest.java

```diff
@@ -0,0 +1,228 @@
+import static junit.framework.TestCase.assertTrue;
+
+import com.google.firebase.firestore.model.protovalue.ObjectValue;
+import com.google.firebase.firestore.model.value.FieldValue;
+import com.google.firestore.v1.Value;
+import java.util.Collections;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(manifest = Config.NONE)
+public class ObjectValueBuilderTest {
+  private Value fooValue = valueOf("foo");
+  private Value barValue = valueOf("bar");
+  private Value emptyObject = valueOf(Collections.emptyMap());
+
+  @Test
+  public void emptyBuilder() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    ObjectValue object = builder.build();
+    assertEquals(wrapObject(), object);
+  }
+
+  @Test
+  public void setSingleField() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    builder.set(field("foo"), fooValue);
+    ObjectValue object = builder.build();
+    assertEquals(wrapObject("foo", fooValue), object);
+  }
+
+  @Test
+  public void setEmptyObject() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    builder.set(field("foo"), emptyObject);
+    ObjectValue object = builder.build();
+    assertEquals(wrapObject("foo", emptyObject), object);
+  }
+
+  @Test
+  public void setMultipleFields() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    builder.set(field("foo"), fooValue);
+    builder.set(field("bar"), fooValue);
+    ObjectValue object = builder.build();
+    assertEquals(wrapObject("foo", fooValue, "bar", fooValue), object);
+  }
+
+  @Test
+  public void setSuccessorField() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    builder.set(field("a"), fooValue);
+    builder.set(field("a0"), fooValue);
+    ObjectValue object = builder.build();
+    assertEquals(wrapObject("a", fooValue, "a0", fooValue), object);
+  }
+
+  @Test
+  public void setNestedField() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    builder.set(field("a.b"), fooValue);
+    builder.set(field("c.d.e"), fooValue);
+    ObjectValue object = builder.build();
+    assertEquals(wrapObject("a", map("b", fooValue), "c", map("d", map("e", fooValue))), object);
+  }
+
+  @Test
+  public void setTwoFieldsInNestedObject() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    builder.set(field("a.b"), fooValue);
+    builder.set(field("a.c"), fooValue);
+    ObjectValue object = builder.build();
+    assertEquals(wrapObject("a", map("b", fooValue, "c", fooValue)), object);
+  }
+
+  @Test
+  public void setFieldInNestedObject() {
+    ObjectValue.Builder builder = ObjectValue.emptyObject().toBuilder();
+    builder.set(field("a"), map("b", fooValue));
+    builder.set(field("a.c"), fooValue);
```

Comment:
Also consider a test where you set a nested object at a more-than-one-level difference of depth.

Reply:
Sure thing. Added `setsDeeplyNestedFieldInNestedObject()`